src/aleph/web/controllers/storage.py

```diff
@@ -56,20 +80,133 @@ async def add_storage_json_controller(request: web.Request):
     return web.json_response(output)
 
 
-async def storage_add_file(request: web.Request):
+async def verify_signature(message: BasePendingMessage) -> bool:
+    """Verifies a signature of a message, return True if verified, false if not"""
+    verification = get_verification_buffer(message)
+
+    message_hash = await run_in_executor(
+        None, functools.partial(encode_defunct, text=verification.decode("utf-8"))
+    )
+
+    verified = False
+    try:
+        # we assume the signature is a valid string
+        address = await run_in_executor(
+            None,
+            functools.partial(
+                Account.recover_message, message_hash, signature=message.signature
+            ),
+        )
+        if address == message.sender:
+            verified = True
+        else:
+            return False
+
+    except Exception as e:
+        verified = False
+    return verified
+
+
+async def get_message_content(
+    post_data: MultiDictProxy[Union[str, bytes, FileField]]
+) -> Tuple[dict, int]:
+    message_bytearray = post_data.get("message", b"")
+    value = post_data.get("size") or 0
+
+    if isinstance(message_bytearray, bytearray):
+        message_string = message_bytearray.decode("utf-8")
+        message_dict = json.loads(message_string)
+        message_dict["time"] = float(message_dict["time"])
+    else:
+        message_dict = {}
+
+    return message_dict, int(str(value))
+
+
+async def verify_and_handle_request(
+    pending_message_db, file_io, message, size, session_factory
+):
```

Comment:
You should split this function in (at least) 3 functions.

1. verify that the message has a valid signature
2. verify that the user has enough tokens to pay for the feature
3. read the file and verify that it matches the item hash.

This will make your code clearer and separate concerns.